Support https://wicg.github.io/webcrypto-secure-curves/#ed25519

youennf opened the pull request:
#### 3abe0e45310ef3e846f8509e44b5514fb2099f76
<pre>
Support <a href="https://wicg.github.io/webcrypto-secure-curves/#ed25519">https://wicg.github.io/webcrypto-secure-curves/#ed25519</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=246145">https://bugs.webkit.org/show_bug.cgi?id=246145</a>
rdar://problem/100401588

Reviewed by Chris Dumez.

Introduce CryptoKeyOKP to support secure safe curves.
Add support for generating, importing and exporting Ed25519 keys.
Add a runtime flag to control exposing the support.

This patch reuses work done in <a href="https://github.com/WebKit/WebKit/pull/5026">https://github.com/WebKit/WebKit/pull/5026</a> by Angela Izquierdo Garcia.

* LayoutTests/http/wpt/crypto/serialize-cryptokey-okp-expected.txt: Added.
* LayoutTests/http/wpt/crypto/serialize-cryptokey-okp.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/sign_verify/eddsa.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/sign_verify/eddsa.https.any.worker-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h: Added.
(cccurve25519_make_priv):
(cccurve25519_make_pub):
(cccurve25519_make_key_pair):
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readOKPKey):
(WebCore::CloneDeserializer::readCryptoKey):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp: Added.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
(WebCore::CryptoAlgorithmEd25519::create):
(WebCore::CryptoAlgorithmEd25519::identifier const):
(WebCore::CryptoAlgorithmEd25519::generateKey):
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
(WebCore::CryptoAlgorithmEd25519::importKey):
(WebCore::CryptoAlgorithmEd25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h: Added.
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp: Added.
(WebCore::keySizeInBytesFromNamedCurve):
(WebCore::CryptoKeyOKP::create):
(WebCore::CryptoKeyOKP::CryptoKeyOKP):
(WebCore::CryptoKeyOKP::generatePair):
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::exportRaw const):
(WebCore::CryptoKeyOKP::exportJwk const):
(WebCore::CryptoKeyOKP::namedCurveString const):
(WebCore::CryptoKeyOKP::isValidOKPAlgorithm):
(WebCore::CryptoKeyOKP::algorithm const):
(WebCore::CryptoKeyOKP::platformSupportedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
(WebCore::CryptoKeyOKP::computePublicKeyFromPrivateKey const):
* Source/WebCore/crypto/keys/CryptoKeyOKP.h: Added.
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Cocoa.cpp: Added.
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/mac/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/mac/CryptoAlgorithmRegistryMac.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp: Added.
(WebCore::CryptoKeyOKP::platformSupportedCurve):
(WebCore::CryptoKeyOKP::computePublicKeyFromPrivateKey const):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
* Source/WebCore/platform/Logging.h:

Canonical link: <a href="https://commits.webkit.org/259277@main">https://commits.webkit.org/259277@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/791b854212f2d105fe150126c6498a73aa4149f7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104512 "14 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37416 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113786 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174010 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108432 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4513 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112748 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38915 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93194 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25994 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80585 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94501 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6946 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27351 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92403 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4730 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7066 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3911 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30025 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13101 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46911 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101089 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6396 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8860 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25088 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt:
```diff
@@ -1,14 +1,14 @@
+PASS Good parameters: Ed25519 bits (jwk, object(kty, crv, x), {name: Ed25519}, true, [])
+PASS Good parameters: Ed25519 bits (raw, buffer(32), {name: Ed25519}, true, [])
+PASS Good parameters: Ed25519 bits (pkcs8, buffer(48), {name: Ed25519}, true, [sign])
+FAIL Good parameters: Ed25519 bits (jwk, object(crv, d, x, kty), {name: Ed25519}, true, [sign]) assert_true: Round trip works expected true got false
```
We're not concerned about this failure?

youennf replied:
Yes, we are. This is documented in the commit message.
There seems to be an issue when deriving the public key from private key, I plan to dig further in a follow-up.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt:
```diff
@@ -1,14 +1,14 @@
+PASS Good parameters: Ed25519 bits (jwk, object(kty, crv, x), {name: Ed25519}, true, [])
+PASS Good parameters: Ed25519 bits (raw, buffer(32), {name: Ed25519}, true, [])
+PASS Good parameters: Ed25519 bits (pkcs8, buffer(48), {name: Ed25519}, true, [sign])
+FAIL Good parameters: Ed25519 bits (jwk, object(crv, d, x, kty), {name: Ed25519}, true, [sign]) assert_true: Round trip works expected true got false
```
ditto.

youennf replied:
Ditto

cdumez commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -95,6 +95,13 @@ static bool isRSAESPKCSWebCryptoDeprecated(JSGlobalObject& state)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(&state);
+    auto* context = globalObject.scriptExecutionContext();
+    return context ? context->settingsValues().webCryptoSafeCurvesEnabled : false;
```
nit: could be:
`return context && context->settingsValues().webCryptoSafeCurvesEnabled;`

youennf replied:
OK

cdumez commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp:
```diff
@@ -0,0 +1,203 @@
+
+CryptoAlgorithmIdentifier CryptoAlgorithmEd25519::identifier() const
+{
+    return s_identifier;
```
nit: Seems this could be inlined in the header?

youennf replied:
ok

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
```diff
@@ -0,0 +1,265 @@
+static constexpr size_t keySizeInBytesFromNamedCurve(CryptoKeyOKP::NamedCurve curve)
+{
+    switch (curve) {
+    case CryptoKeyOKP::NamedCurve::X25519:
```
Can merge with the next `case`, no need for another `return` statement.

youennf replied:
done

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
```diff
@@ -0,0 +1,265 @@
+{
+    switch (m_curve) {
+    case NamedCurve::X25519:
+        return String(X25519);
```
nit: You probably don't need the `String()`

youennf replied:
done

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyOKP.cpp:
```diff
@@ -0,0 +1,265 @@
+    case NamedCurve::X25519:
+        return String(X25519);
+    case NamedCurve::Ed25519:
+        return String(Ed25519);
```
ditto.

youennf replied:
done

cdumez commented on Source/WebCore/crypto/keys/CryptoKeyOKP.h:
```diff
@@ -0,0 +1,92 @@
+    String generateJwkD() const;
+    String generateJwkX() const;
+
+    static bool platformSupportedCurve(NamedCurve);
```
I think this should be named `isPlatformSupportedCurve()` given that this returns a boolean.

youennf replied:
done

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
```diff
@@ -0,0 +1,326 @@
+            return { };
+        }
+    }
+    auto publicKey = CryptoKeyOKP::create(identifier, namedCurve, CryptoKeyType::Public, Vector<uint8_t>(ccPublicKey), true, usages);
```
Passing `true` here is not very readable, we should use an enum class.

youennf replied:
I used a local variable to make that readable

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
```diff
@@ -0,0 +1,326 @@
+    result.append(101);
+    switch (namedCurve()) {
+        case NamedCurve::X25519:
+            result.append(110);
```
We may want to use global constants for these things. Especially given that they are used in several places (and need to match).

youennf replied:
done

cdumez commented on Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
```diff
@@ -0,0 +1,326 @@
+            result.append(110);
+            break;
+        case NamedCurve::Ed25519:
+            result.append(112);
```
ditto.

youennf replied:
done

cdumez reviewed (approved):
LGTM

youennf committed:
Support https://wicg.github.io/webcrypto-secure-curves/#ed25519
https://bugs.webkit.org/show_bug.cgi?id=246145
rdar://problem/100401588

Reviewed by Chris Dumez.

Introduce CryptoKeyOKP to support secure safe curves.
Add support for generating, importing and exporting Ed25519 keys.
Add a runtime flag to control exposing the support.

This patch reuses work done in https://github.com/WebKit/WebKit/pull/5026 by Angela Izquierdo Garcia.

* LayoutTests/http/wpt/crypto/serialize-cryptokey-okp-expected.txt: Added.
* LayoutTests/http/wpt/crypto/serialize-cryptokey-okp.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey_failures_Ed25519.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/sign_verify/eddsa.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/sign_verify/eddsa.https.any.worker-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h: Added.
(cccurve25519_make_priv):
(cccurve25519_make_pub):
(cccurve25519_make_key_pair):
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::write):
(WebCore::CloneDeserializer::read):
(WebCore::CloneDeserializer::readOKPKey):
(WebCore::CloneDeserializer::readCryptoKey):
* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::normalizeCryptoAlgorithmParameters):
(WebCore::isSupportedExportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.cpp: Added.
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
(WebCore::CryptoAlgorithmEd25519::create):
(WebCore::CryptoAlgorithmEd25519::identifier const):
(WebCore::CryptoAlgorithmEd25519::generateKey):
(WebCore::CryptoAlgorithmEd25519::sign):
(WebCore::CryptoAlgorithmEd25519::verify):
(WebCore::CryptoAlgorithmEd25519::importKey):
(WebCore::CryptoAlgorithmEd25519::exportKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmEd25519.h: Added.
* Source/WebCore/crypto/keys/CryptoKeyOKP.cpp: Added.
(WebCore::keySizeInBytesFromNamedCurve):
(WebCore::CryptoKeyOKP::create):
(WebCore::CryptoKeyOKP::CryptoKeyOKP):
(WebCore::CryptoKeyOKP::generatePair):
(WebCore::CryptoKeyOKP::importRaw):
(WebCore::CryptoKeyOKP::importJwk):
(WebCore::CryptoKeyOKP::exportRaw const):
(WebCore::CryptoKeyOKP::exportJwk const):
(WebCore::CryptoKeyOKP::namedCurveString const):
(WebCore::CryptoKeyOKP::isValidOKPAlgorithm):
(WebCore::CryptoKeyOKP::algorithm const):
(WebCore::CryptoKeyOKP::platformSupportedCurve):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
(WebCore::CryptoKeyOKP::computePublicKeyFromPrivateKey const):
* Source/WebCore/crypto/keys/CryptoKeyOKP.h: Added.
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Cocoa.cpp: Added.
(WebCore::signEd25519):
(WebCore::verifyEd25519):
(WebCore::CryptoAlgorithmEd25519::platformSign):
(WebCore::CryptoAlgorithmEd25519::platformVerify):
* Source/WebCore/crypto/mac/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/mac/CryptoAlgorithmRegistryMac.cpp:
(WebCore::CryptoAlgorithmRegistry::platformRegisterAlgorithms):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp: Added.
(WebCore::CryptoKeyOKP::platformSupportedCurve):
(WebCore::CryptoKeyOKP::computePublicKeyFromPrivateKey const):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
* Source/WebCore/platform/Logging.h:

Canonical link: https://commits.webkit.org/259277@main
